README.md
Plemp!
======

Create your own online pile of junk!  Plemp allows you to upload/put stuff
on a single canvas, for sharing, collaging or any purpose you can think of.

Features
--------

* Adding content by means of typing/pasting text
* Uploading of image/music/video/code/text files
* (Re)arranging and resizing of the items
* Some visual effects for eye candy purposes
* Syntax highlighting of source code files
* Distributed management of the arrangement, i.e. everybody viewing the
  Plemp page will (with some delay) see the same arrangement

Requirements
------------

Plemp! is an Express (on node.js) application, so you need:

* Node.js (>= 0.4.12) with
  - Express (>= 2.5.0)
  - Connect-Form (>= 0.2.0)
* JQuery (>= 1.7.0)
* JQuery UI (>= 1.8.16)

Installation
------------

For now, Plemp! is in a developing state and not ready for site-wide
deployment yet.  However, before running, make sure that jquery.js,
jquery-ui.js, and jquery-ui.css with images are available from public/
either by copying or symlinking them there.

Usage
-----

Run from the command line:

    $ node plemp.js

and head over to http://localhost:3300/ to view and use the Plemp! canvas.

Files can be uploaded or text can be pasted using the Add/Upload Dialog
invoked by pressing the Plus key or the Plus button in the top-right
corner.  When your browser supports it, files can also be uploaded by
dragging it and dropping it on the canvas.
Once uploaded, the objects can be arranged by dragging them around and
resized by using the resize handles.

License
-------

Plemp! is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

*N.B.* The source code of this program includes code from external projects.

The following files are taken from the highlight.js library (see
http://softwaremaniacs.org/soft/highlight/en/):

* public/javascripts/highlight.pack.js
* public/stylesheets/highlight.css

These files are licensed under the 3-clause BSD license:
https://github.com/isagalaev/highlight.js/blob/master/LICENSE.

The following file is the Jeditable jQuery plugin (see
http://www.appelsiini.net/projects/jeditable):

* public/javascript/jquery-jeditable.min.js

This file is licensed under the MIT license:
http://www.opensource.org/licenses/mit-license.php.

The following file is the jQuery filedrop plugin (see
https://github.com/weixiyen/jquery-filedrop):

* public/javascript/jquery-filedrop.js

This file is licensed under the MIT license:
http://www.opensource.org/licenses/mit-license.php.


NEWS
Plemp! News
===========

Version 0.4
-----------

The last missing feature with respect to the original Plemp! is added:

* Support multiple clients at once that can see each other's changes in
  (almost) real-time

New features are:

* Support for uploads via drag & drop
* Style tweaks
* Fixed: escape HTML symbols in text
* Fixed: properly set titles from the form
* Many other small fixes

Version 0.2
-----------

Implemented many features that where still missing with respect to the
original version of Plemp!

* Allow for the deletion of items
* Support for music/video/code next to normal text files
* Keyboard support (Plus/Escape) for showing/hiding the upload dialog

New features are:

* Raise items on click
* Tweaked dragging feel
* Added place holder for comments
* Make titles editable
* Syntax highlighting


Version 0.1
-----------

Initial version; rewrite based on node.js.


db.js
var fs = require("fs")
  , path = require("path")
  , db_file = "./draggables.json";

// Saves the draggables info object to the database file (synchronously).
function save(obj) {
  console.log("Saved the database to " + db_file);
  console.dir(obj);
  fs.writeFileSync(db_file, JSON.stringify(obj));
}

// Loads the draggables info object from the database file.
function load() {
  if (path.existsSync(db_file)) {
    var json_string = fs.readFileSync(db_file);
  }
  else {
    // If the database file does not exist, return a new, empty object.
    return {};
  }

  try {
    var obj = JSON.parse(json_string);
    console.log("Loaded the database from " + db_file);
    console.dir(obj);
    return obj;
  }
  catch (exception) {
    console.error("There was an error parsing the database file: ", exception);
    process.exit(1);
  }
}

exports.save = save
exports.load = load


app.js
// plemp.rb - The Plemp! application, create your own on-line pile of junk!
//
// Plemp! is Copyright © 2012 Paul van Tilburg <paul@mozcode.nl>
//
// This program is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation; either version 2 of the License, or (at your option) any later
// version.

var express = require("express")
  , db = require("./db")
  , form = require("connect-form")
  , fs = require("fs")
  , path = require("path")
  , crypto = require("crypto")
  , mime = require("mime")
  , url = require("url");

// Set up the Node Express application.
var app = express.createServer(form({ keepExtensions: true,
                                      uploadDir: __dirname + '/public/upload' }));

// List of events.
var events = [];
// List of deferred requests/connections.
var defers = [];
// Maximum age of events (in seconds).
var maxAge = 3600;
// Request sequence number.
var lastRequestId = 0;

// Return the current time time in milliseconds.
function currentTimestamp() {
  return new Date().getTime();
}

// Escape the HTML.
function escapeHTML(text) {
  return text.replace(/&/g, '&amp;')
             .replace(/</g, '&lt;')
             .replace(/>/g, '&gt;')
             .replace(/"/g, '&quot;')
             .replace(/'/g, '&#039;');
}

// Compacts an array by removing all undefined values.
function compact(arr) {
  if (!arr) return null;

  var i, data = [];
  for (i = 0; i < arr.length; i++) {
    if (arr[i]) data.push(arr[i]);
  }
  return data;
}

// Add a new event with the given type and optional data.
function addEvent(type, data) {
  var event = { type: type,
                timestamp: currentTimestamp() };
  if (data) event.data = data;

  events.push(event);
  console.log("[P] " + JSON.stringify(event));
  // Notify deferred connections that there is something new.
  notify();
}

// Find the next event in the list of events after the optional timestamp.
// Expires events that are older than the maximum age (maxAge).
function nextEvent(timestamp) {
  if (!events) return null;
  if (!timestamp) timestamp = 0;

  var event, nEvent, i;
  var minTimestamp = currentTimestamp() - maxAge * 1000;
  for (i = 0; i < events.length; i++) {
    event = events[i];

    // Check if event is expired.
    if (event.timestamp < minTimestamp) {
      console.log("[.] expired: " + JSON.stringify(event));
      delete events[i];
      continue;
    }
    // Check if event is newer.
    if (event.timestamp > timestamp) {
      console.log("[.] next event after timestamp " + timestamp + ": " +
                  JSON.stringify(event));
      nEvent = event;
      break;
    }
  }
  // Compact the list of events.
  events = compact(events);
  return nEvent;
}

// Notify all deferred connections of their respect nextEvent.
function notify() {
  if (!defers) return;

  var i, ctx, event;
  for (i = 0; i < defers.length; i++) {
    ctx = defers[i];

    if (!ctx.req) {
      // Apparently this connectioned was timed out.
      delete defers[i];
      continue;
    }

    event = nextEvent(ctx.timestamp);
    if (event) {
      ctx.req.resume();
      ctx.res.send(event);
      ctx.res.end();
      delete defers[i];
      console.log("[" + ctx.id + "] sent " + JSON.stringify(event));
    }
  }
  // Compact the list of deferred connections.
  defers = compact(defers);
}

// Save and pause the request.
function pause(timestamp, req, res, requestId) {
  var ctx = { id: requestId,
              timestamp: timestamp,
              req: req,
              res: res };
  defers.push(ctx);

  req.connection.setTimeout(60 * 1000);
  req.connection.on('timeout', function() {
    ctx.req = null;
    ctx.res = null;
    console.log("[" + requestId + "] timed out");
  });

  req.pause();
  console.log("[" + requestId + "] paused");
}

// Retrieve the draggables info.
var draggables = db.load();
for (drag_id in draggables) {
  var drag = draggables[drag_id];

  if (!path.existsSync(__dirname + "/public/upload/" + drag.name)) {
    console.log("Could not find file for draggable " + drag_id +
                "; removing from database!");
    delete draggables[drag_id];
  }
}

// Application settings and middleware configuration.
app.configure(function() {
  app.use(express.logger());
  app.use(express.static(__dirname + '/public'));
  app.use(express.bodyParser());
  app.use(express.methodOverride());
  app.use(app.router);
});

app.configure('development', function() {
  app.use(express.errorHandler({ dumpExceptions: true, showStack: true }));
});

app.configure('production', function() {
  app.use(express.errorHandler());
});

// Server the main index file statically for now.
app.get('/', function(req, res) {
  res.redirect('/index.html');
});

// The events controller: accessed through AJAX long-poll requests by the
// main page for getting new events.
app.get('/events', function(req, res) {
  var u = url.parse(req.url, true);

  if (!u.query) {
    res.send(null, 400);
    return;
  }

  var timestamp = u.query.timestamp || 0,
      requestId = lastRequestId++,
      event = nextEvent(timestamp);

  if (!event) {
    pause(timestamp, req, res, requestId);
  }
  else {
    res.send(event);
    res.end();
    console.log("[" + requestId + "] direct sent " + JSON.stringify(event));
  }
});

// The download controller: provides access to the uploaded files but ensures
// that they are downloaded by the client.
app.get('/download/:id', function(req, res) {
  var drag_id = req.params.id;
  var drag = draggables[drag_id];
  console.log("Provide download for draggable " + drag_id);
  res.download(__dirname + "/public/upload/" + drag.name);
});

// The retrieval controller: accessed through AJAX requests by the main
// page for getting/setting the state (positions) of the draggables.
app.get('/draggables', function(req, res) {
  // Retrieve the current status of the draggables and return in JSON format.
  res.send({ timestamp: currentTimestamp(),
             list: draggables });
});

// The upload controller: handles uploads from the main site.  This can
// either be a file or some pasted text.  After upload the controler
// redirects to the main page which includes the just uploaded file.
app.post('/draggables', function(req, res, next) {
  req.form.complete(function(err, fields, files) {
    if (err) {
      next(err);
    }
    else {
      var drag_id, file_name, file_mime, file_title;
      if (fields.text) {
        var md5sum = crypto.createHash('md5');
        md5sum = md5sum.update(fields.text).digest('hex');
        drag_id = md5sum
        file_name = drag_id + "." + fields.type;
        file_title = fields.title ||  "Untitled";
        file_mime = mime.lookup(fields.type);
        fs.writeFile(__dirname + "/public/upload/" + file_name,
                     fields.text, function (err) {
          if (err)
            throw err;
          console.log('Text saved to %s', file_name);
        });
        // FIXME: prevent this file being created from the start!
        fs.unlink(files.file.path);
      }
      else {
        console.log('File %s uploaded to %s', files.file.filename,
                                              files.file.path);
        file_name = path.basename(files.file.path);
        drag_id = path.basename(file_name, path.extname(file_name));
        file_title = fields.title ||
                     path.basename(files.file.filename,
                                   path.extname(files.file.filename));
        file_mime = files.file.mime || "application/octet-stream";
      }
      draggables[drag_id] = { name: file_name,
                              mime: file_mime,
                              title: file_title,
                              top: 200,
                              left: 350,
                              width: 400,
                              height: 300 };
      addEvent("add", { id: drag_id, info: draggables[drag_id] });
      // If this was a drag & drop upload, do not redirect to /.
      if (fields.type == 'dnd') {
        res.send(true);
        res.end();
      }
      else {
        res.redirect('home');
      }
    }
  });
});

// The draggable controller: provides direct access to the HTML
// generating code for draggable objects.
app.get('/draggables/:id', function(req, res) {
  var drag_id = req.params.id;
  var drag = draggables[drag_id];
  var file_name = "../upload/" + drag.name;
  console.log("Get draggable: " + drag_id);
  // Stuff taken from the Camping implementation.
  var default_style = "left:" + drag.left + "px;top:" + drag.top + "px;" +
                      "width:" + drag.width + "px;height:" + drag.height + "px;" +
                      "display:none;";
  var title = drag.title || drag.name || 'Untitled';
  var content;
  var mime_type = drag.mime.split("/");
  switch (mime_type[0]) {
    case "image":
      content = '<img src="' + file_name + '"></img>';
      break;
    case "video":
      content = '<video src="' + file_name + '" controls>' +
                '<p>Your browser does not support HTML5 video.</p>' +
                '</video>';
      break;
    case "audio":
      content = '<audio src="' + file_name + '" controls>' +
                '<p>Your browser does not support HTML5 video.</p>' +
                '</audio>'
      break;
    case "text":
      file_contents = fs.readFileSync(__dirname + "/public/upload/" + drag.name, 'utf8');
      content = '<pre>' + escapeHTML(file_contents) + '</pre>';
      break;
    case "application": // FIXME: treat as code for now, but it is probably wrong
      file_contents = fs.readFileSync(__dirname + "/public/upload/" + drag.name, 'utf8');
      content = '<pre><code class="' + drag.type + '">' +
                     escapeHTML(file_contents) +
                     '</code></pre>';
      break;
    default:
      content = '<span>Unknown type: ' + mime_type + '</span>';
  }

  // Wrap the content in a div with title and comments.
  res.send('<div class="draggable" id="' + drag_id + '" ' +
                'style="' + default_style + '">' +
             '<h2><span class="title">' + title + '</span>' +
                  '<div class="delete" title="Delete…">&#10799;</div>' +
                  '<div class="download" title="Download…">&#8595;</div>' +
             '</h2>' + content + '<div class="comments">Comments (0)</div>' +
           '</div>');
});

// The position save controller: access through AJAX request by the main
// page for committing position changes of the draggables to the database,
// i.e. the global state.
app.post('/draggables/:id', function(req, res) {
  var drag = draggables[req.params.id];
  if (!drag) {
    res.send(false);
    return;
  }
  if (req.body.title) {
    // It's a title update!
    console.log("Title update for draggable " + req.params.id + ": " +
                req.body.title);
    drag.title = req.body.title;
    addEvent("title update", { id: req.params.id, title: req.body.title });
    res.send(req.body.title);
  }
  else {
    if (req.body.width) {
      // Set the size for the file with the given ID.
      console.log("Size update for draggable " + req.params.id + ";" +
                  " width: " + req.body.width +
                  " height: " + req.body.top);
      drag.width = req.body.width;
      drag.height = req.body.height;
      addEvent("resize", { id: req.params.id, width: drag.width, height: drag.height });
    }
    else {
      // Set the position for the file with the given ID.
      console.log("Position update for draggable " + req.params.id + ";" +
                  " left: " + req.body.left +
                  " top: " + req.body.top);
      drag.top = req.body.top;
      drag.left = req.body.left;
      addEvent("reposition", { id: req.params.id, top: drag.top, left: drag.left });
    }
  }
});

// Draggable removal controller: removes the specific draggable from the
// database.
app.del('/draggables/:id', function(req, res) {
  var drag_id = req.params.id;
  var drag = draggables[drag_id];
  fs.unlink(__dirname + "/public/upload/" + drag.name, function(err) {
    if (err) {
      console.log("Something went wrong while deleting draggable " +
                  drag_id + ": " + err);
      res.send(err);
      throw err;
      return;
    }
    delete draggables[drag_id];
    console.log("Deleted draggable " + drag_id);
    addEvent("delete", { id: req.params.id });
    res.send(true);
  });
});

// Signal handling.
process.on('SIGINT', function() { db.save(draggables); process.exit(0); });
process.on('SIGTERM', function() { db.save(draggables); process.exit(0); });

// Start the application.
app.listen(3300);
console.log('Plemp! started on http://127.0.0.1:%d/ in %s mode',
            app.address().port, app.settings.env);


plemp.js
// plemp.rb - The Plemp! application, create your own on-line pile of junk!
//
// Plemp! is Copyright © 2012 Paul van Tilburg <paul@mozcode.nl>
//
// This program is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation; either version 2 of the License, or (at your option) any later
// version.

var express = require("express")
  , db = require("./db")
  , form = require("connect-form")
  , fs = require('fs')
  , path = require('path')
  , crypto = require("crypto")
  , mime = require("mime")

// Set up the Node Express application.
var app = express.createServer(form({ keepExtensions: true,
                                      uploadDir: __dirname + '/public/upload' }));

// Retrieve the draggables info.
var draggables = db.load();
for (drag_id in draggables) {
  if (!path.existsSync(__dirname + "/public/upload/" + drag_id)) {
    console.log("Could not find draggable " + drag_id +
                "; removing from database!");
    delete draggables[drag_id];
  }
}

// Application settings and middleware configuration.
app.configure(function() {
  app.use(express.logger());
  app.use(express.static(__dirname + '/public'));
  app.use(express.bodyParser());
  app.use(express.methodOverride());
  app.use(app.router);
  app.use(express.errorHandler({ dumpExceptions: true, showStack: true }));
});

// Server the main index file statically for now.
app.get('/', function(req, res) {
  res.redirect('/index.html');
});

// The retrieval controller: accessed through AJAX requests by the main
// page for getting/setting the state (positions) of the draggables.
app.get('/draggables', function(req, res) {
  // Retrieve the current status of the draggables and return in JSON format.
  res.send(draggables);
});

// The upload controller: handles uploads from the main site.  This can
// either be a file or some pasted text.  After upload the controler
// redirects to the main page which includes the just uploaded file.
app.post('/draggables', function(req, res) {
  req.form.complete(function(err, fields, files) {
    if (err) {
      next(err);
    }
    else {
      var file_id, file_name, file_mime;
      if (fields.text) {
        md5sum = crypto.createHash('md5');
        md5sum = md5sum.update(fields.text).digest('hex')
        file_id = md5sum + "." + fields.type
        file_name = "public/upload/" + file_id;
        file_mime = mime.lookup(fields.type);
        fs.writeFile(file_name, fields.text, function (err) {
          if (err)
            throw err;
          console.log('Text saved to %s', file_name);
        });
        // FIXME: prevent this file being created from the start!
        fs.unlink(files.file.path);
      }
      else {
        console.log('File %s uploaded to %s', files.file.filename,
                                              files.file.path);
        file_id = path.basename(files.file.path);
        file_name = files.file.filename;
        file_mime = files.file.mime;
      }
      draggables[file_id] = { name: files.file.filename,
                              mime: file_mime,
                              title: fields.title,
                              top: 200,
                              left: 350 };
    }
  });
  res.redirect('home');
});

// The draggable controller: provides direct access to the HTML
// generating code for draggable objects.
app.get('/draggables/:id', function(req, res) {
  var drag_id = req.params.id;
  var file_name = "../upload/" + req.params.id;
  console.log("Get draggable: " + drag_id);
  // Stuff taken from the Camping implementation.
  var drag = draggables[drag_id];
  var default_style = "left:" + drag.left + "px;top:" + drag.top + "px;";
  var title = drag.title || drag.name || 'Title not set';
  var content;
  var mime_type = drag.mime.split("/")
  switch (mime_type[0]) {
    case "image":
      content = '<img src="' + file_name + '"></img>';
      break;
    case "video":
      content = '<video src="' + file_name + '" controls="true"></video>';
      break;
    case "audio":
      content = '<audio src="' + file_name + '" controls="true"></audio>';
      break;
    case "text":
      file_contents = fs.readFileSync("public/upload/" + drag_id);
      content = '<pre>' + file_contents + '</pre>';
      break;
    case 'application': // FIXME: treat as code for now, but it is probably wrong
      file_contents = fs.readFileSync("public/upload/" + drag_id);
      content = '<pre><code class="' + drag.type + '">' + file_contents +
                     '</code></pre>';
      break;
    default:
      content = '<span>Unknown type: ' + mime_type + '</span>';
  }

  // Wrap the content in a div with title and comments.
  res.send('<div class="draggable" id="' + drag_id + '" ' +
                'style="' + default_style + '">' +
             '<h2><span class="title">' + title + '</span>' +
                  '<div class="delete">X</div>' +
             '</h2>' + content + '<div class="comments"></div>' +
           '</div>');
});

// The position save controller: access through AJAX request by the main
// page for committing position changes of the draggables to the database,
// i.e. the global state.
app.post('/draggables/:id', function(req, res) {
  var drag = draggables[req.params.id];
  if (req.body.title) {
    // It's a title update!
    console.log("Title update for draggable " + req.params.id + ": " +
                req.body.title);
    drag.title = req.body.title;
    res.send(req.body.title);
  }
  else {
    // Set the position for the file with the given ID.
    console.log("Position update for draggable " + req.params.id + ";" +
                " left: " + req.body.left +
                " top: " + req.body.top);
    drag.top = req.body.top;
    drag.left = req.body.left;
  }
});

// Draggable removal controller: removes the specific draggable from the
// database.
app.del('/draggables/:id', function(req, res) {
  var file_id = req.params.id;
  console.log("Hello?");
  fs.unlink("public/upload/" + file_id, function(err) {
    if (err) {
      console.log("Something went wrong while deleting " +
                  file_id + ": " + err);
      res.send(err);
      throw err;
    }
    delete draggables[file_id];
    console.log("Deleted draggable " + file_id);
    res.send(true);
  });
});

// Signal handling.
process.on('SIGINT', function() { db.save(draggables); process.exit(0); });
process.on('SIGTERM', function() { db.save(draggables); process.exit(0); });

// Start the application.
app.listen(3300);
console.log('Plemp! started on http://127.0.0.1:%d/', app.address().port)


public/javascripts/plemp-ui.js
$(document).ready(function() {
  // Handlers for showing/hiding the upload dialog via clicks & keys.
  $("#add").click(show_add_dialog);
  $("#add_form #cancel").click(hide_add_dialog);
  if ($.browser.mozilla) {
    $(document).keypress(key_handler);
  }
  else {
    $(document).keyup(key_handler);
  }

  // The plempable jQuery plugin containing most of the Plemp! UI
  // functionality.
  $.fn.plempable = function(data) {
    var type = data.mime.split("/")[0];
    this.draggable({ stack: ".draggable",
                     containment: "window",
                     cancel: "audio, video",
                     distance: 10,
                     stop: update_drag_info })
        .click(raise_draggable);
    // Make it resizable with parameters based on the type:
    // FIXME: is there a better way than repeating the common options?
    switch(type) {
      case "image":
      case "video":
        this.resizable({ distance: 10,
                         start: raise_draggable,
                         stop: update_drag_info,
                         handles: "se",
                         minHeight: 250,
                         aspectRatio: true });
        break;
      case "audio":
        this.resizable({ distance: 10,
                         start: raise_draggable,
                         stop: update_drag_info,
                         handles: "e",
                         minWidth: 300 });
        break;
      case "text":
      case "application":
        this.resizable({ distance: 10,
                         start: raise_draggable,
                         stop: update_drag_info,
                         handles: "se",
                         minWidth: 300,
                         minHeight: 150,
                         alsoResize: "#" + this.attr("id") + " pre" });
    }
    // Configure events for child elements.
    this.find(".download").click({ id: this.attr("id") }, download_draggable);
    this.find(".delete").click({ id: this.attr("id") }, confirm_delete_draggable);
    this.find(".title").editable('draggables/' + this.attr("id"),
                                 { tooltip: "Click to edit…",
                                   name: 'title',
                                   submit: 'OK',
                                   cancel: 'Cancel',
                                   style: 'inherit' });
    // FIXME: actually get the comments!
    //this.find(".comments").text("Comments (0)");
    // Fix the height of the pre block.  FIXME: can this be improved?
    this.find("pre").css({ height: (this.height() - 50) + "px" });
    // Highlight contained code in the pre block.
    this.find("pre code").each(function(idx, elem) {
      hljs.highlightBlock(elem, '  ');
    });
    return $(this);
  };

  // Set up the dialogs.
  $("#add_dialog").dialog({ autoOpen: false,
                            width: "auto",
                            show: "fade",
                            hide: "fade",
                            resizable: false,
                            modal: true,
                            close: clear_add_dialog_form,
                            buttons: { "Cancel":  hide_add_dialog,
                                       "Upload!": function() {
                                         $("#add_form")[0].submit();
                                       }}});
  $("#delete_confirm").dialog({ autoOpen: false,
                                show: "fade",
                                hide: "fade",
                                resizable: false,
                                buttons: { "Cancel": function() {
                                             $(this).dialog("close");
                                           },
                                           "Delete": function() {
                                             $(this).dialog("close");
                                             delete_draggable($(this).data("id"));
                                           }}});

  // For drag & drop file uploading via the filedrop jQuery plugin and HTML5.
  $("body").filedrop({ url: '/draggables',
                       paramname: 'file',
                       data: { type: 'dnd' },
                       maxfiles: 1,
                       maxfilesize: 50,
                       error: handle_filedrop_error
                     });

  // Populate the canvas with the draggables.
  $.get("draggables", function(data) {
    var timestamp = data.timestamp;

    // Retrieve each draggable from the list and add it to the canvas.
    $.each(data.list, function(drag_id, drag_info) {
      add_draggable_to_canvas(drag_id, drag_info);
    });

    // Start a (long-poll) loop for update events from the server.
    poll_server(timestamp);
  }, "json");
});

// Callback functions

// Clear the form of the add dialog.
function clear_add_dialog_form() {
  $("#add_form")[0].reset();
}

// Show the add dialog with a visual effect.
function show_add_dialog() {
  $("#add_dialog").dialog("open");
}

// Hide the add dialog with a visual effect and reset the form when finished.
function hide_add_dialog() {
  $("#add_dialog").dialog("close");
}

// Add a draggable element to the canvas.
function add_draggable_to_canvas(drag_id, drag_info) {
  $.get("draggables/" + drag_id, function(data) {
    $("#draggables").append(data);
    // Assume we have appended a signle draggable here:
    $(".draggable").last().plempable(drag_info).fadeIn('slow');
  });
}

// Delete a draggable element from the canvas.
function delete_draggable_from_canvas(drag_id) {
  $("#draggables #" + drag_id).hide('fade', 'slow', function() {
    $(this).remove();
  });
}

// Confirm the deletion of the draggable using the confirm dialog.
function confirm_delete_draggable(event) {
  event.preventDefault();
  $("#delete_confirm").data("id", event.data.id)
                      .dialog("open");
}

// Deletion has been confirmed, delete the draggable from the server and
// then the canvas.
function delete_draggable(drag_id) {
  $.post("draggables/" + drag_id, {"_method": "delete"}, function(data) {
    if (data) delete_draggable_from_canvas(drag_id);
  }, "json");
}

// Download a draggable from the server.
function download_draggable(event) {
  event.preventDefault();
  window.open("download/" + event.data.id);
}

// Raise the draggable to the foreground.
function raise_draggable(event) {
  var max = 0;
  $(".draggable").each(function() {
    var cur = parseFloat($(this).css("z-index"));
    max = cur > max ? cur : max;
  });
  $(this).css("z-index", max + 1);
}

// Update the position of a draggable on the server.
function update_drag_info(event, ui) {
  if (ui.size) {
    $.post("draggables/" + ui.helper.context.id, ui.size, "json");
  }
  else {
    $.post("draggables/" + ui.helper.context.id, ui.position, "json");
  }
}

// Poll for server events via AJAX.
function poll_server(timestamp) {
  $.ajax({ url: "events?timestamp=" + timestamp,
           success: handle_server_event,
           error: function(event) {
             switch (event.statusText) {
               case "timeout":
                 // Got a timeout, retry with the same timestamp.
                 poll_server(timestamp);
                 break;
               case "abort":
               case "error":
                 // Lost connection to the server for some reason!
                 set_message("Lost contact to the server somehow! " +
                             "Reload this page to try to fix that.");
            }},
           dataType: "json", timeout: 60 * 1000 });
};

// Handle server events.
function handle_server_event(event) {
  var drag = $("#draggables #" + event.data.id);
  switch(event.type) {
    case "reposition":
      drag.animate({ top: event.data.top, left: event.data.left });
      break;
    case "resize":
      drag.animate({ width: event.data.width, height: event.data.height })
          // Fix the height of the pre.  FIXME: can this be improved?
          .find("pre").animate({ height: event.data.height - 50 });
      break;
    case "title update":
      $("#draggables #" + event.data.id + " h2 .title").text(event.data.title);
      break;
    case "add":
      add_draggable_to_canvas(event.data.id, event.data.info);
      break;
    case "delete":
      delete_draggable_from_canvas(event.data.id);
      break;
    default:
      console.log("Unsupported event type!");
      console.log(event);
  }
  poll_server(event.timestamp);
}

// Handle filedrop (drag & drop uploading) errors.
function handle_filedrop_error(err, file) {
  switch(err) {
    case 'BrowserNotSupported':
        flash_message("Your browser does not support drag & drop! " +
                       "Please click this button to upload:");
        break;
    case 'TooManyFiles':
        // user uploaded more than 'maxfiles'
        flash_message("You can only upload a single file at once!");
        break;
    case 'FileTooLarge':
        // program encountered a file whose size is greater than 'maxfilesize'
        // FileTooLarge also has access to the file which was too large
        // use file.name to reference the filename of the culprit file
        flash_message("The file is too large! " +
                      "Ensure it is smaller than than 50MB.");
        break;
    default:
        console.log(err, file);
        break;
  }
}

// Flash some messsage.
function flash_message(msg) {
  $("#message").html(msg)
  $("#message").fadeIn('slow').delay(5000).fadeOut('fast');
}

// Set some messsage (permanent version of flash_message).
function set_message(msg) {
  $("#message").html(msg)
  $("#message").fadeIn('slow');
}

// Handle the Escape and Plus keys for hiding/showing the add dialog.
function key_handler(event) {
  switch (event.keyCode) {
    case 27: /* Escape */
      hide_add_dialog();
      break;
    case 187: /* Plus (numpad) */
      show_add_dialog();
      break;
  }
}
